allennlp/models/segmental_language_model.py

```diff
@@ -0,0 +1,300 @@
+        return_dict['projection'] = projected_bi
+
+        # compute softmax loss
+        # TODO(Swabha): What does embeddings do for loss computation?
+        forward_loss, backward_loss = self._compute_loss(projected_bi,
+                                                        embeddings,
+                                                        forward_targets,
+                                                        backward_targets)
+
+        num_targets = torch.sum((forward_targets > 0).float())
+        if num_targets > 0:
+            if self._bidirectional:
+                total_loss = 0.5 * (forward_loss + backward_loss)
+                average_loss = total_loss / num_targets
+            else:
+                total_loss = forward_loss
+                average_loss = forward_loss / num_targets
+            return_dict.update({
+                    'loss': average_loss,
+                    'forward_loss': forward_loss / num_targets,
+                    'backward_loss': (backward_loss / num_targets
+                                        if backward_loss is not None else None),
+                    'batch_weight': num_targets
+            })
+        else:
+            # average_loss zero tensor, return it for all
+            total_loss = average_loss = torch.tensor(0.0).to(forward_targets.device)  # pylint: disable=not-callable
+            return_dict.update({
+                    'loss': average_loss,
+                    'forward_loss': average_loss,
+                    'backward_loss': average_loss if backward_loss is not None else None
+            })
+
+        # this is stored to compute perplexity if needed
+        self._last_average_loss[0] = average_loss.detach().item()
+        # Send metrics for evaluation.
+        self.metric(loss=total_loss, num_targets=num_targets)
+
+        return return_dict
+
+    def _get_segmental_embeddings(self,
+                                  encoder: Seq2SeqEncoder,
+                                  unidirectional_embs: torch.tensor,
+                                  boundaries: torch.LongTensor,
+                                  mapping: torch.LongTensor):
+        # First, collect the hidden states of segment boundaries.
+        seg_boundary_embs, seg_boundary_mask = self._get_gathered_embeddings(
+            embeddings=unidirectional_embs,
+            indices=boundaries)
+
+        # Pass through forward or backward encoder.
+        seg_embeddings = encoder(seg_boundary_embs, seg_boundary_mask)
+        seg_embeddings_with_dropout = self._dropout(seg_embeddings)
+
+        # Secondly, the segmental embeddings need to be scattered, so each
+        # position gets its own segmental information.
+        seg_embeddings_scattered, _ = self._get_gathered_embeddings(
+            embeddings=seg_embeddings_with_dropout,
+            indices=mapping)
+        return seg_embeddings_scattered
+
+    @staticmethod
+    def _get_gathered_embeddings(embeddings: torch.Tensor,
```

Comment:
Could you add some more comments for these last two methods? What they do at a high level, basically. I was having a bit of trouble following.